File: backend/src/routes/simulate.js
'use strict';
/**
 * POST   /simulate/batch  — generate N demo closed trades
 * DELETE /simulate/reset  — wipe all trades, params, and adaptation logs
 */
const { Router } = require('express');
const crud                = require('../db/crud');
const { DEFAULT_PARAMS }  = require('../engine/parameters');
const { runAdaptation }   = require('../engine/adaptation');
const { Trade, ParameterVersion, AdaptationLog, sequelize } = require('../db/models');

const router  = Router();
const SYMBOLS = ['EURUSD', 'GBPUSD', 'USDJPY', 'XAUUSD'];

/** Gaussian random using Box-Muller */
function _gauss(mean = 0, std = 1) {
  const u = 1 - Math.random();
  const v = Math.random();
  return mean + std * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}

function _brownianPrice(base = 1.085, steps = 1, volatility = 0.0005) {
  const prices = [base];
  for (let i = 0; i < steps; i++) {
    prices.push(Math.round((prices[prices.length - 1] + _gauss(0, volatility)) * 100000) / 100000);
  }
  return prices;
}

router.post('/simulate/batch', async (req, res) => {
  const count      = Math.min(Math.max(parseInt(req.query.count      || '20',   10), 1),  200);
  const winRatePct = Math.min(Math.max(parseFloat(req.query.win_rate_pct || '52'), 0), 100);

  let params = await crud.getCurrentParams();
  if (!params) {
    await crud.saveParams(DEFAULT_PARAMS, 'Initial defaults', 'SYSTEM');
    params = { ...DEFAULT_PARAMS };
  }

  const latestPv = await ParameterVersion.findOne({ order: [['version', 'DESC']] });
  const version  = latestPv ? latestPv.version : 1;

  let basePrice = 1.08500;
  const tradeIds = [];

  for (let i = 0; i < count; i++) {
    const symbol    = SYMBOLS[Math.floor(Math.random() * SYMBOLS.length)];
    const direction = Math.random() > 0.5 ? 'BUY' : 'SELL';
    const price     = _brownianPrice(basePrice, 1)[1];
    basePrice       = price;

    const atr    = Math.random() * (0.0025 - 0.0005) + 0.0005;
    const slDist = price * (params.stop_loss_pct  / 100);
    const tpDist = price * (params.take_profit_pct / 100);

    const sl = direction === 'BUY'
      ? Math.round((price - slDist) * 100000) / 100000
      : Math.round((price + slDist) * 100000) / 100000;
    const tp = direction === 'BUY'
      ? Math.round((price + tpDist) * 100000) / 100000
      : Math.round((price - tpDist) * 100000) / 100000;

    const isWin = Math.random() * 100 < winRatePct;
    const exitPrice = isWin ? tp : sl;
    const pnl = isWin
      ?  Math.round(Math.abs(tp - price) * params.lot_size * 100000 * 100) / 100
      : -Math.round(Math.abs(sl - price) * params.lot_size * 100000 * 100) / 100;
    const result = isWin ? 'WIN' : 'LOSS';

    const now         = Date.now();
    const offsetMs    = Math.floor(Math.random() * 1440 + 5) * 60000;
    const openedAt    = new Date(now - offsetMs);
    const durationMs  = Math.floor(Math.random() * (240 - 10) + 10) * 60000;
    const closedAt    = new Date(openedAt.getTime() + durationMs);
    const durationMin = durationMs / 60000;

    const trade = await crud.logTrade({
      symbol, direction,
      entry_price:      price,
      exit_price:       exitPrice,
      stop_loss:        sl,
      take_profit:      tp,
      lot_size:         params.lot_size,
      pnl,
      result,
      duration_mins:    Math.round(durationMin * 10) / 10,
      atr_at_entry:     Math.round(atr * 100000) / 100000,
      ema_fast_at_entry: Math.round((price - atr * 0.3) * 100000) / 100000,
      ema_slow_at_entry: Math.round((price - atr * 0.7) * 100000) / 100000,
      params_version:   version,
      opened_at:        openedAt,
      closed_at:        closedAt,
    });

    // Patch timestamps directly (Sequelize already wrote them via logTrade above)
    await Trade.update(
      { opened_at: openedAt, closed_at: closedAt },
      { where: { id: trade.id } },
    );

    tradeIds.push(trade.id);
  }

  const adaptResult = await runAdaptation(Math.min(count, 20));

  return res.json({
    simulated:  count,
    trade_ids:  tradeIds,
    adaptation: adaptResult,
    stats:      await crud.getStats(),
  });
});

router.delete('/simulate/reset', async (_req, res) => {
  await AdaptationLog.destroy({ where: {} });
  await ParameterVersion.destroy({ where: {} });
  await Trade.destroy({ where: {} });
  return res.json({ status: 'reset complete' });
});

module.exports = router;

File: python-backend/app/routers/simulate.py
import random
from datetime import datetime, timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy import delete, desc, select
from sqlalchemy.orm import Session

from .. import crud
from ..db import get_db
from ..models import AdaptationLog, ParameterVersion, Trade
from ..services.adaptation import run_adaptation
from ..strategy.dtc import DEFAULT_PARAMS, compute_levels, resolve_params

router = APIRouter()
SYMBOLS = ["EURUSD", "GBPUSD", "USDJPY", "XAUUSD"]


@router.post("/simulate/batch")
def simulate_batch(
    count: int = Query(20, ge=1, le=200),
    win_rate_pct: float = Query(52, ge=0, le=100),
    db: Session = Depends(get_db),
):
    params_raw = crud.get_current_params(db)
    if not params_raw:
        crud.save_params(db, DEFAULT_PARAMS.copy(), reason="Initial defaults", trigger="SYSTEM")
        params_raw = DEFAULT_PARAMS.copy()
    params = resolve_params(params_raw)
    latest = db.scalar(select(ParameterVersion).order_by(desc(ParameterVersion.version)).limit(1))
    version = latest.version if latest else 1

    base_price = 1.08500
    trade_ids = []
    for _ in range(count):
        symbol = random.choice(SYMBOLS)
        direction = "BUY" if random.random() > 0.5 else "SELL"
        price = round(base_price + random.uniform(-0.0020, 0.0020), 5)
        base_price = price
        levels = compute_levels(direction, price, params)
        is_win = random.random() * 100 < win_rate_pct
        exit_price = levels["tp1"] if is_win else levels["sl"]
        pnl = (abs(exit_price - price) * params.lot_size * 100000) * (1 if is_win else -1)
        opened_at = datetime.utcnow() - timedelta(minutes=random.randint(5, 1200))
        closed_at = opened_at + timedelta(minutes=random.randint(10, 240))
        trade = crud.log_trade(
            db,
            {
                "symbol": symbol,
                "direction": direction,
                "entry_price": price,
                "exit_price": exit_price,
                "stop_loss": levels["sl"],
                "take_profit": levels["tp1"],
                "lot_size": params.lot_size,
                "pnl": round(pnl, 2),
                "result": "WIN" if is_win else "LOSS",
                "duration_mins": round((closed_at - opened_at).total_seconds() / 60.0, 1),
                "atr_at_entry": round(random.uniform(0.0005, 0.0025), 5),
                "ema_fast_at_entry": round(price - random.uniform(0.0001, 0.0008), 5),
                "ema_slow_at_entry": round(price - random.uniform(0.0003, 0.0012), 5),
                "params_version": version,
                "opened_at": opened_at,
                "closed_at": closed_at,
            },
        )
        trade_ids.append(trade.id)

    adapt_result = run_adaptation(db, window=min(count, settings_for_window()))
    return {"simulated": count, "trade_ids": trade_ids, "adaptation": adapt_result, "stats": crud.get_stats(db)}


def settings_for_window() -> int:
    return 20


@router.delete("/simulate/reset")
def simulate_reset(db: Session = Depends(get_db)):
    db.execute(delete(AdaptationLog))
    db.execute(delete(ParameterVersion))
    db.execute(delete(Trade))
    db.commit()
    return {"status": "reset complete"}
